### ISSUE TITLE
Closures formatted strangely

### ISSUE THREAD

**Issue by marcusklaas** [rust-lang-nursery/rustfmt#470]
``` rust
fn main() {
    {{{
        let explicit_arg_decls =
            explicit_arguments.into_iter()
                              .enumerate()
                              .map(|(index, (ty, pattern))| {
                                  let lvalue = Lvalue::Arg(index as u32);
                                  block = this.pattern(block,
                                                       argument_extent,
                                                       hair::PatternRef::Hair(pattern),
                                                       &lvalue);
                                  ArgDecl { ty: ty }
                              });
    }}}
}
```

becomes

``` rust
fn main() {
    {
        {
            {
                let explicit_arg_decls =
                    explicit_arguments.into_iter()
                                      .enumerate()
                                      .map(|(index, (ty, pattern))| {
                                              let lvalue = Lvalue::Arg(index as u32);
                                              block = this.pattern(block,
                                                                   argument_extent,
                                                                   hair::PatternRef::Hair(pattern),
                                                                   &lvalue);
                                              ArgDecl { ty: ty }
                                          });
            }
        }
    }
}
```

and after another pass of rustfmt becomes

``` rust
fn main() {
    {
        {
            {
                let explicit_arg_decls = explicit_arguments.into_iter()
                                                           .enumerate()
                                                           .map(|(index, (ty, pattern))| {
                                                               let lvalue =
                                                                   Lvalue::Arg(index as u32);
                                                               block = this.pattern(block,
                                                                   argument_extent,
                                                                   hair::PatternRef::Hair(pattern),
                                                                   &lvalue);
                                                               ArgDecl { ty: ty }
                                                           });
            }
        }
    }
}
```

.
_Super_ weird.

**Comment by nrc** [comment]
bug for non-idempotence

**Comment by marcusklaas** [comment]
We're running into the limits of `FmtVisitor` here. When it fails to format something, it just takes its span and moves on. Instead, we should abort. The `visit::Visitor<'v>` trait doesn't support this.

I think we should just move block formatting to the same model as that of expression formatting. The most significant change would probably involve missing spans. We will no longer be able to write them directly to the string buffer.

**Comment by nrc** [comment]
This no longer reproduces as described, in particular the result is now idempotent. However, the result is still bad, the arguments to `this.pattern` are poorly aligned:

```
fn main() {
    {
        {
            {
                let explicit_arg_decls = explicit_arguments.into_iter()
                                                           .enumerate()
                                                           .map(|(index, (ty, pattern))| {
                                                               let lvalue =
                                                                   Lvalue::Arg(index as u32);
                                                               block = this.pattern(block,
                                                       argument_extent,
                                                       hair::PatternRef::Hair(pattern),
                                                       &lvalue);
                                                               ArgDecl { ty: ty }
                                                           });
            }
        }
    }
}
```

**Comment by nrc** [comment]
Closed by #1502
